### PR TITLE
Add diags for non-existing binary used by prebuild plugin

### PR DESCRIPTION
**Pull request by elsh**
Add diags for non-existing binary used by prebuild plugin.
Add a test to verify a package with a plugin target no longer requires a non-plugin target.
Resolves rdar://90442392, rdar://86787091

### CODE REVIEW

**Comment by elsh** [comment]
@swift-ci smoke test

File: Sources/SPMBuildCore/PluginInvocation.swift [hunk 1]
```diff
@@ -188,6 +188,10 @@ extension PluginTarget {
                         outputFiles: try outputFiles.map{ try AbsolutePath(validating: $0) })
                     
                 case .definePrebuildCommand(let config, let outputFilesDir):
+                    let execPath = try AbsolutePath(validating: config.executable)
+                    if !FileManager.default.fileExists(atPath: execPath.pathString) {
```
Comment (tomerd) [inline Sources/SPMBuildCore/PluginInvocation.swift:192]: use our file system abstraction?

File: Tests/SPMBuildCoreTests/PluginInvocationTests.swift [hunk 2]
```diff
@@ -579,4 +579,162 @@ class PluginInvocationTests: XCTestCase {
...
+            // Create a plugin script runner for the duration of the test.
+            let pluginCacheDir = tmpPath.appending(component: "plugin-cache")
+            let pluginScriptRunner = DefaultPluginScriptRunner(
+                fileSystem: localFileSystem,
+                cacheDir: pluginCacheDir,
+                toolchain: try UserToolchain.default
+            )
+
+            // Define a plugin compilation delegate that just captures the passed information.
+            class Delegate: PluginScriptCompilerDelegate {
```
Comment (tomerd) [inline Tests/SPMBuildCoreTests/PluginInvocationTests.swift:698]: used?

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

File: Sources/SPMBuildCore/PluginInvocation.swift [hunk 3]
```diff
@@ -188,6 +188,10 @@ extension PluginTarget {
                         outputFiles: try outputFiles.map{ try AbsolutePath(validating: $0) })
                     
                 case .definePrebuildCommand(let config, let outputFilesDir):
+                    let execPath = try AbsolutePath(validating: config.executable)
+                    if !localFileSystem.exists(execPath) {
```
Comment (tomerd) [inline Sources/SPMBuildCore/PluginInvocation.swift:192]: hard coding the local file system safe? or is it pluggable?

**Comment by elsh** [comment]
@swift-ci  smoke test

File: Sources/SPMBuildCore/PluginInvocation.swift [hunk 4]
```diff
@@ -383,14 +382,16 @@ extension PackageGraph {
...
                     let delegateQueue: DispatchQueue
                     let toolPaths: [AbsolutePath]
+                    let fileSystem: FileSystem
                     var outputData = Data()
                     var diagnostics = [Basics.Diagnostic]()
                     var buildCommands = [BuildToolPluginInvocationResult.BuildCommand]()
                     var prebuildCommands = [BuildToolPluginInvocationResult.PrebuildCommand]()
                     
-                    init(delegateQueue: DispatchQueue, toolPaths: [AbsolutePath]) {
+                    init(delegateQueue: DispatchQueue, toolPaths: [AbsolutePath], fileSystem: FileSystem) {
```
Comment (tomerd) [inline Sources/SPMBuildCore/PluginInvocation.swift:391]: nit: make `fileSystem` the first argument (we usually so). otherwise looks great!

**Review by tomerd — COMMENTED** [review]
one small nit, otherwise looks great!

**Review by tomerd — APPROVED** [review]

File: Sources/SPMBuildCore/PluginInvocation.swift (hunk 4 above)
Comment (elsh) [reply Sources/SPMBuildCore/PluginInvocation.swift:391]: Is there a reason why it's made to be the first argument in general?

**Comment by elsh** [comment]
@swift-ci smoke test

**Review by neonichu — APPROVED** [review]

**Pull request merged**